meetings/2017-03-03.md

```diff
@@ -0,0 +1,352 @@
+Hoped to have more concrete proposal but we don’t. @mhdawson: Should we open a
+public repo with current work and to continue conversation? @ofrobots: Perhaps
+in a couple months, once we have a concrete proposal. @jasnell: Even a list of
+open issues would be helpful.
+
+@ofrobots: As module authors migrate to NAPI, some may be able to avoid C++
+entirely and choose FFI instead.
+
+@mrkmarron: Note that a strong FFI implementation exposes Node.js to the risk of
+users taking more dependencies on legacy native code. That could interfere with
+tools that help with diagnostics and optimizations.
+
+Concern about error handling, for example, when a flow calls back and forth from
+JS to native several times; exceptions might not propagate through multiple
+back-and-forth calls.
+
+@mcollina: Some native modules maintain state on the native side referenceable
+from JS. How would that be addressed with current proposal? nanomsg in fact has
+this issue, how did you handle it? @matthewloring: Did the best we could, would
+be good to think about this more.
+
+@mcollina: Simplest case is a synchronous call passing a couple primitives back
+and forth. That’s useful but doesn’t address more complex modules which for
+example may need access to native objects. @matthewloring: Considered providing
+pointers within JS.
+
+@matthewloring: FFI can provide strongly opinionated flows for common scenarios.
+N-API addresses flows which require different opinions.
+
+@jasnell: Since native APIs for FFI would be included in NAPI, other opinionated
+flows could be developed too.
+
+Timeline? Too early to say, not V8 5.7 or 5.8 so not the first release of Node
+8.
+
+Will it be possible to backport? @ofrobots: Probably not to 6.
+
+Timeline for providing a public drop for collaboration? Part that depends on VM
+implementation needs more maturity before sharing.
+
+Can work on a shared type definition format now and utilize that for early
+implementations.
+
+**Next steps**
+
+* Could begin discussing a shared type definition format now.
+* Matt and Ali to continue work on FFI implementation in private, no timeline
+  for public drop.
+
+---
+
+# N-API
+
+Presentation:
+<https://github.com/nodejs/abi-stable-node/blob/doc/VM%20Summit.pdf>
+
+Error handling: NAPI captures errors and exceptions and module author can handle
+as desired. Also possible to use `napi_throw` functions to throw an exception in
+JS.
+
+C++ API wrapper over C API for convenience, but not planning to include in core.
+
+V8 Isolate APIs little used in modules and don’t plan to include in N-API.
+[deasync](https://www.npmjs.com/package/deasync) uses many of these.
+
+Demo from Arunesh: https://github.com/boingoing/napi_demo. Same native modules -
+canvas, leveldown - used with 6, 8, and Node-ChakraCore.
+
+Canvas perf regression is likely because there’s heavy usage of V8 APIs so much
+more marshaling needed. Also first tried with C++ wrapper. No optimizations yet
+either.
```

Comment:
I suggest

Canvas perf regression is likely because the performance benchmark makes calls to canvas where the work done in the native method is very small. This means that any overhead, even the limited overhead added by N-API has a significant impact.